src/Sentry/Internal/Http/CachingTransport.cs

```diff
@@ -278,9 +281,24 @@ private async Task ProcessCacheAsync(CancellationToken cancellation)
                 await InnerProcessCacheAsync(file, cancellation).ConfigureAwait(false);
             }
 
+            if (Interlocked.Exchange(ref _transportIsFailing, 0) == 1)
+            {
+                // Retry envelopes that got stuck in processing while the transport was failing
+                MoveUnprocessedFilesBackToCache();
+                _workerSignal.Release();
+            }
+
             // Signal that we can continue with initialization, if we're using _options.InitCacheFlushTimeout
             _initCacheResetEvent?.Set();
         }
+        catch (Exception e)
```

Comment:
> Should we add a `when` clause here? Otherwise, we'd potentially run into a situation where a corrupted envelope gets moved between `processing` and `cache` back and forth?

Yeah, I think we need some mechanism to prevent envelopes that are failing for non-network related reasons from getting stuck in an endless loop. @bruno-garcia [mentioned that](https://github.com/getsentry/sentry-dotnet/issues/80#issuecomment-728352819) in the original discussion around building the `CachingTransport`. See more detailed discussion below...

> But that still does not fix the broken retry-logic we have on 168, right?

By 168 you the task delay here?
https://github.com/getsentry/sentry-dotnet/blob/d2d074bb08ed7c354224136eded29228d3a6057b/src/Sentry/Internal/Http/CachingTransport.cs#L161-L169

I'm not entirely sure why a delay was added here to be honest. The loop won't run again until the `_workerSignal` is released, which (prior to this PR) only happened when an envelope was queued to be sent:
https://github.com/getsentry/sentry-dotnet/blob/d2d074bb08ed7c354224136eded29228d3a6057b/src/Sentry/Internal/Http/CachingTransport.cs#L466

In this PR I tweaked it so that **_if_** the transport failed to send an envelope **_and subsequently_** the transport was able to send some other envelope (a fairly imprecise mechanism for detecting when a network issue might have manifested and then subsequently been resolved) then we ressurrect any envelopes that might have been stuck in processing purgatory and have another go at sending them.

## Comments / Suggested Improvements

### Detecting network errors

We need a better way to detect when envelopes fail to send due to network issues. At the moment, if we have one "bad" envelope that fails to send because because of something freaky about the way it serializes, followed by good envelopes forever then that bad envelope would currently go back and forth between processing and cache forever (just taking up non-disk system resources unnecessarily). 

Maybe, as you suggest, we could use the `Exception.Type` to infer this.

### Detecting network recovery

The trigger for resending an envelope is currently that another envelope is successfully sent (since implicitly, if we were able to send some other envelope, the network must be back online).

A better solution might be some kind of progressive backoff/retry timer to check periodically whether the network is back online... that way the resend would be guaranteed even if no other envelopes get enqueued after the network connection recovers.

### Dealing with non-network errors

Assuming we find a reliable mechanism to detect when there is a network issue and also to detect when the network reconnects again, currently when "retrying" we just copy all of the files that are stuck in processing back into the cache. If 97 of those failed due to non-network related reasons and 3 of them failed due to network related reasons, there's no reason to try to send all 100 of them again. Ideally we'd only resend the envelopes that failed because the network was unavailable when we last tried to send them. 

I think we need a list of those envelopes that we think deserve a second chance, vs those that failed for other reasons.... 

Those that failed for other reasons should probably go in some kind of dead letter queue. I don't see how we can automatically recover from other kinds of failure - these really require human intervention. So I think that might be the ultimate solution to that problem.

Reply:
So I finally realised there was a bunch of logic already in the `InnerProcessCacheAsync` method that models out most of what we want. [In our main branch](https://github.com/getsentry/sentry-dotnet/blob/main/src/Sentry/Internal/Http/CachingTransport.cs#L291-L360) (i.e. prior to this PR) that catches any operation cancelled exceptions and re-throws these, then it catches any network exceptions and logs these (but doesn't delete the envelope)... finally it catches any other exceptions, logs these and deletes the envelope. 

So the desired behaviour for failing to send an envelope appears to be simply to discard the envelope. I rolled back plans to create a dead letter queue then as I don't think this is how we want the SDK to behave... we wouldn't ever have any plan to be able to successfully send these messages so that makes sense. 

All I'm doing now then is flagging when we run into a network error and then using that flag after sending an envelope successfully to detect when a previously unavailable network is available again. When that happens, we try to resend any envelopes that failed to send due to the network being unavailable. 

Given the broader logic for the `InnerProcessCacheAsync` method, we know we'll only be resending envelopes that failed due to network connectivity issues (other envelopes would have been deleted). 

So I think this is a pretty good solution. 

The only small improvement that might be made is to throw some kind of timer in there to check, periodically, for network connectivity and resend envelopes as soon as the network comes back on line again... rather than waiting for some other envelope to be sent successfully (which is currently now the network coming back on line gets detected). That feels more like a nice to have than a need to have though.

Reply:
I'm SO looking forward digging deeper there! Thanks for all the sleuthing!

Reply:
> I'm SO looking forward digging deeper there! Thanks for all the sleuthing!

You probably want to grab the sample code in the issue description to test this - I've basically been firing up that console app and opening the cache directory in Finder... then playing around with turning the wifi on/off. Functionally, everything should be working now. What people think of my coding style is another matter 😜